From 49b7fa5e1c04d3c9481db6c96982515455f16a1a Mon Sep 17 00:00:00 2001
From: King Jin <jinking.this@gmail.com>
Date: Fri, 25 Aug 2023 16:46:29 +0800
Subject: [PATCH] add --prerequires to support Debian Pre-Depends

---
 checkinstall-man.sgml |  7 +++++++
 checkinstall.in       | 21 ++++++++++++++++++++-
 2 files changed, 27 insertions(+), 1 deletion(-)

--- a/checkinstall-man.sgml
+++ b/checkinstall-man.sgml
@@ -249,6 +249,13 @@
       </varlistentry>
 
       <varlistentry>
+        <term><option>--prerequires</option>
+        </term>
+	<listitem><para>Dependencies that must be installed and configured before this package.</para>
+        </listitem>
+      </varlistentry>
+
+      <varlistentry>
         <term><option>--rpmflags</option>
         </term>
 	<listitem><para>Pass this flags to the rpm installer.</para>
--- a/checkinstall.in
+++ b/checkinstall.in
@@ -151,6 +151,7 @@
    echog "--maintainer=<email addr>      The package maintainer (.deb)"
    echog "--provides=<list>              Features provided by this package"
    echog "--requires=<list>              Features required by this package"
+   echog "--prerequires=<list>           Packages that must be installed and configured before this package"
    echog "--recommends=<list>            Features recommended by this package"
    echog "--suggests=<list>              Features suggested by this package"
    echog "--conflicts=<list>             Packages that this package cannot be installed with (.deb)"
@@ -514,7 +515,7 @@
 # Arguments parsing
 
 CKNAME=`basename "$0"`
-PARAMS=`getopt -a -n $CKNAME -o +d:DA:t:RShHy -l arch:,type:,si,showinstall::,ss,showslack::,deldoc::,delspec::,deldesc::,strip::,addso::,install::,stripso::,gzman::,bk,backup::,autodoinst::,reset-uids::,fstrans::,spec:,exclude:,include:,pkgname:,pkgversion:,pkgrelease:,pkglicense:,pkggroup:,pkgsource:,pkgaltsource:,pakdir:,docdir:,requires:,recommends:,suggests:,provides:,conflicts:,replaces:,maintainer:,dpkgflags:,rpmflags:,pkgarch:,umask:,with-tar:,summary:,inspect,review-spec,review-control,newslack,help,nodoc,rpmi,rpmu,version,copyright,default -- "$@"`
+PARAMS=`getopt -a -n $CKNAME -o +d:DA:t:RShHy -l arch:,type:,si,showinstall::,ss,showslack::,deldoc::,delspec::,deldesc::,strip::,addso::,install::,stripso::,gzman::,bk,backup::,autodoinst::,reset-uids::,fstrans::,spec:,exclude:,include:,pkgname:,pkgversion:,pkgrelease:,pkglicense:,pkggroup:,pkgsource:,pkgaltsource:,pakdir:,docdir:,requires:,prerequires:,recommends:,suggests:,provides:,conflicts:,replaces:,maintainer:,dpkgflags:,rpmflags:,pkgarch:,umask:,with-tar:,summary:,inspect,review-spec,review-control,newslack,help,nodoc,rpmi,rpmu,version,copyright,default -- "$@"`
 
 [ $? -gt 0 ] && help_notice
 
@@ -605,6 +606,10 @@
          shift
          REQUIRES=`eval echo $1`
          ;;
+      --prerequires)
+         shift
+         PREREQUIRES=`eval echo $1`
+         ;;
       --recommends)
          shift
          RECOMMENDS=`eval echo $1`
@@ -1282,8 +1287,11 @@
    PKG_GROUP=`getparameter "^[Gg]roup"`
    PROVIDES=`getparameter "^[Pp]rovides"`
    REQUIRES=`getparameter "^[Rr]equires"`
+   PREREQUIRES=`getparameter "^[Pp]rerequires"`
    RECOMMENDS=`getparameter "^[Rr]commends"`
    SUGGESTS=`getparameter "^[Ss]uggests"`
+   CONFLICTS=`getparameter "^[Cc]onflicts"`
+   REPLACES=`getparameter "^[Rr]eplaces"`
 fi
       
    OPTION=junk
@@ -1382,8 +1390,11 @@
       ! [ "$ALTSOURCE" ] && ALTSOURCE=""
       ! [ "$PROVIDES" ] && PROVIDES="$NAME"
       ! [ "$REQUIRES" ] && REQUIRES=""
+      ! [ "$PREREQUIRES" ] && PREREQUIRES=""
       ! [ "$RECOMMENDS" ] && RECOMMENDS=""
       ! [ "$SUGGESTS" ] && SUGGESTS=""
+      ! [ "$CONFLICTS" ] && CONFLICTS=""
+      ! [ "$REPLACES" ] && REPLACES=""
       # bond: added this so it is easy to change the Maintainer: field
       # just by setting the MAINTAINER environment variable
 
@@ -1427,6 +1438,7 @@
           echog "13 - Provides: [ %s ]" "$PROVIDES"
       [ "$CK_DEBIAN" ] && echog "14 - Conflicts: [ %s ]" "$CONFLICTS"
       [ "$CK_DEBIAN" ] && echog "15 - Replaces: [ %s ]" "$REPLACES"
+      [ "$CK_DEBIAN" ] && echog "16 - Prerequires: [ %s ]" "$PREREQUIRES"
 
       echo
       echogn "Enter a number to change any of them or press ENTER to continue: "
@@ -1531,6 +1543,12 @@
             read REPLACES
             }
             ;;
+        16)
+            [ "$CK_DEBIAN" ] && { echog "Enter the pre-required packages: "
+            echo -n ">> "
+            read PREREQUIRES
+            }
+            ;;
       esac
    done
    
@@ -2623,6 +2641,7 @@
 Architecture: $ARCHITECTURE
 Version: ${VERSION}-${RELEASE}
 Depends: $REQUIRES
+Pre-Depends: $PREREQUIRES
 Recommends: $RECOMMENDS
 Suggests: $SUGGESTS
 Provides: $PROVIDES
